src/platform/Darwin/BleConnectionDelegateImpl.mm

```diff
@@ -559,33 +571,29 @@ - (void)start
 
 - (void)stop
 {
+    dispatch_assert_queue_debug(_workQueue);
     _scannerDelegate = nil;
     _found = false;
     [self stopScanning];
     [self removePeripheralsFromCache];
 
-    if (!_centralManager && !_peripheral) {
-        return;
+    // Call connection error completion in case we were still in the process of connecting
+    [self dispatchConnectionError:BLE_ERROR_APP_CLOSED_CONNECTION];
```

Comment:
So this is a behavior change for the following cases:

1. NewConnection with some appState followed by another NewConnection for the same appState before the first thing is done.  We _used_ to just drop the old thing on the floor (which was presumably bad if the appState was different!), but now we will report a connection error which the caller will think refers to the _new_ NewConnection attempt.  I wonder whether we should handle this by skipping the error dispatch if the appState matches the old one in NewConnection....
2. `CancelConnection` followed by `NewConnection`.  This one could be bad for two reasons.  First, a cancel can in fact indicate that the appState is going away, so queueing a notification involving appstate from there is not good.  Second, the `NewConnection` can end up using the same appState as the thing being canceled, in which case this is like case 1.  In practice, it looks like `CancelConnection` can come from two places: `StopScan` (which has nothing to dispatch an error to anyway) and `CancelBleIncompleteConnection`.  That last is called from `SetUpCodePairer::StopConnectOverBle` which definitely does _not_ expect an error callback from doing that. It's also called from shutdown bits in `BLEBase::ClearState`, but I would think the setup code pairer would have gotten shut down before then...  Anyway, the upshot is that it feels like in the `CancelConnection` case we shoud not dispatch error. And we really need way better API contracts for all this stuff, so that would be clearer.  :(

Reply:
> StopScan
StopScan should probably not call CancelConnection anyway... the assumption that you can't even scan and connect at the same time doesn't seem to be expressed in the API anywhere, and goes beyond the "1 connection" assumption.

> NewConnection with some appState followed by another NewConnection for the same appState
It's hard to even consider that a valid sequence of calls, if there is no intervening `CancelConnection`...

Reply:
> StopScan should probably not call CancelConnection anyway..

Yes, agreed.

> It's hard to even consider that a valid sequence of calls

The question is whether we end up doing it in practice from CHIPDeviceController/SetUpCodePairer.  :(

Reply:
I pared this down to a minimal crash-fix (hopefully) for now, I'll tackle the rest in a more substantial refactor as a separate PR.